bundles/org.openhab.binding.solarman/src/main/java/org/openhab/binding/solarman/internal/util/StreamUtils.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2010-2024 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.solarman.internal.util;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.function.BiFunction;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * @author Catalin Sanda - Initial contribution
+ */
+@NonNullByDefault
+public class StreamUtils {
+    public static <A, B, C> Stream<C> zip(Stream<? extends A> a, Stream<? extends B> b,
+            BiFunction<? super A, ? super B, ? extends C> zipper) {
+        Objects.requireNonNull(zipper);
+        Spliterator<? extends A> aSpliterator = Objects.requireNonNull(a).spliterator();
+        Spliterator<? extends B> bSpliterator = Objects.requireNonNull(b).spliterator();
+
+        // Zipping looses DISTINCT and SORTED characteristics
+        int characteristics = aSpliterator.characteristics() & bSpliterator.characteristics()
+                & ~(Spliterator.DISTINCT | Spliterator.SORTED);
+
+        long zipSize = ((characteristics & Spliterator.SIZED) != 0)
+                ? Math.min(aSpliterator.getExactSizeIfKnown(), bSpliterator.getExactSizeIfKnown())
+                : -1;
+
+        Iterator<A> aIterator = Spliterators.iterator(aSpliterator);
+        Iterator<B> bIterator = Spliterators.iterator(bSpliterator);
+        Iterator<C> cIterator = new Iterator<C>() {
+            @Override
+            public boolean hasNext() {
+                return aIterator.hasNext() && bIterator.hasNext();
+            }
+
+            @Override
+            public C next() {
+                return zipper.apply(aIterator.next(), bIterator.next());
+            }
+        };
+
+        Spliterator<C> split = Spliterators.spliterator(cIterator, zipSize, characteristics);
+        return (a.isParallel() || b.isParallel()) ? StreamSupport.stream(split, true)
+                : StreamSupport.stream(split, false);
+    }
+
+    public static Collection<Object> reverse(List<Integer> list) {
```

Comment:
Can;t you make use of the default java api? As reverse methods are available:
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collections.html

Reply:
The default API reverses the list in place, whenever possible I like to use APIs without side-effects. Unless this is a blocker, I'll stick with this method.

Reply:
Might not be a blocke,r but trrying to udnerstand. What is the actual side effect?

Reply:
The side effect is that the original list is modified directly when using Collections.reverse. Even though the original list isn't used elsewhere in this specific instance, I prefer using pure functions whenever possible. Pure functions, which avoid side effects, make the code more predictable and easier to reason about. They improve testability and reduce the likelihood of bugs caused by unintended modifications to data. This approach aligns with best practices in functional programming, promoting safer and more maintainable code. I hope this clarifies my reasoning.